R/path_linters.R

```diff
@@ -0,0 +1,208 @@
+win32_good_punct_regex <- rex(one_of("!", "#", "$", "%", "&", "'", "(", ")", "+", ",", "-", ".", ";",
+                                     "=", "@", "[", "]", "^", "_", "`", "{", "}", "~"))
+# win32BadPunct + win32AllowedPunct = [:punct:]
+
+unsafe_char_regex <- rex(or(control_char_regex, win32_bad_punct_regex))
+
+safe_char_regex <- rex(or(alnum, " ", win32_good_punct_regex))
+# alphanumeric chars (including UTF-8), spaces, and win32-allowed punctuation
+
+portable_char_regex <- rex(character_class("A-Za-z0-9_\\-."))
+# ASCII letters, digits, dot, hyphen and underscore
+
+protocol_regex <- rex(one_or_more(letter), "://")
+
+root_regex <- list(
+  unix  = rex(start, "/"),
+  tilde = rex(start, "~", zero_or_more(portable_char_regex), zero_or_more("/")),
+  win32 = rex(start, letter, ":", maybe(one_of("/", "\\"))),
+  unc   = rex(start, "\\\\", maybe(one_or_more(portable_char_regex), maybe("\\")))
+)
+
+root_path_regex <- rex(start, or(root_regex), end)
+
+absolute_path_regex <- rex(or(
+  root_regex[["unix"]]  %if_next_isnt% one_of(space, "/"),
+  root_regex[["tilde"]] %if_next_isnt% one_of(space, quote),
+  root_regex[["win32"]] %if_next_isnt% one_of(space, "/", "\\"),
+  root_regex[["unc"]]   %if_next_isnt% one_of(space, "\\", quote)
+))
+
+relative_path_regex <- rex(
+  start %if_next_isnt% or(quote, absolute_path_regex, protocol_regex),
+  or(                                                  # not an absolute path or protocol, and then
+    group(one_or_more(safe_char_regex), or("/", "\\")),  # (back)slash-separated paths
+    group(dot, maybe(dot), end)                        # or single or double dot
+  )                                                    # (paths without slash or dots left out)
+  %if_next_isnt% quote
+)
+
+path_regex <- rex(or(absolute_path_regex, relative_path_regex))
+
+is_absolute_path <- function(str) {
+  re_matches(str, absolute_path_regex)
+}
+
+is_root_path <- function(str) {
+  re_matches(str, root_path_regex)
+}
+
+is_relative_path <- function(str) {
+  re_matches(str, relative_path_regex)
+}
+
+is_path <- function(str) {
+  re_matches(str, path_regex)
+}
+
+is_valid_path <- function(str, lax=FALSE) {
+  # Given a character vector of paths, return FALSE for directory or file having valid characters.
+  # On Windows, invalid chars are all control chars and: * ? " < > | :
+  # On Unix, all characters are valid, except when lax=TRUE (use same invalid chars as Windows).
+  as.logical(
+    Map(
+      function(dirs, is_win32) {
+        if (is_win32 || lax) {
+          if (length(dirs) && is_root_path(dirs[[1L]])) {
+            dirs <- tail(dirs, -1L)  # remove root element ("/", "C:", or "\\")
+          }
+          !any(re_matches(dirs, unsafe_char_regex))
+        } else {
+          TRUE  # either Unix path or strict (lax=FALSE)
+        }
+      },
+      split_path(str),
+      re_matches(str, rex(or(list(root_regex[["win32"]], root_regex[["unc"]], "\\"))))
+    )
+  )
+}
+
+is_long_path <- function(str) {
```

Comment:
The three functions `is_valid_path()`, `is_long_path()` and `is_valid_long_path()` do indeed expect some sort of path as input. It makes sense to change the argument name as you suggested.

I am not so convinced for changing the argument name for the four functions `is_absolute_path()`, `is_root_path()`, `is_relative_path()`, `is_path()` because the input is a string and the functions determine if this is a path.